sklearn/decomposition/tests/test_pca.py

```diff
@@ -192,141 +118,126 @@ def test_whitening():
+    pca_other = PCA(n_components=2, svd_solver=svd_solver, random_state=0)
+
+    pca_full.fit(X)
+    pca_other.fit(X)
+
+    assert_allclose(
+        pca_full.explained_variance_,
+        pca_other.explained_variance_,
+        atol=1e-2
+    )
+    assert_allclose(
+        pca_full.explained_variance_ratio_,
+        pca_other.explained_variance_ratio_,
+        atol=1e-4
+    )
+
+
+@pytest.mark.parametrize(
+    'X',
+    [np.random.RandomState(0).randn(100, 80),
+     datasets.make_classification(100, 80, n_informative=78,
+                                  random_state=0)[0]],
+    ids=['random-data', 'correlated-data']
+)
+@pytest.mark.parametrize('svd_solver', PCA_SOLVERS)
+def test_pca_explained_variance_empirical(X, svd_solver):
+    pca = PCA(n_components=2, svd_solver=svd_solver, random_state=0)
+    X_pca = pca.fit_transform(X)
+    assert_allclose(pca.explained_variance_, np.var(X_pca, ddof=1, axis=0))
 
-    # compare to empirical variances
     expected_result = np.linalg.eig(np.cov(X, rowvar=False))[0]
     expected_result = sorted(expected_result, reverse=True)[:2]
+    assert_allclose(pca.explained_variance_, expected_result, atol=1e-2)
 
-    X_pca = pca.transform(X)
-    assert_array_almost_equal(pca.explained_variance_,
-                              np.var(X_pca, ddof=1, axis=0))
-    assert_array_almost_equal(pca.explained_variance_, expected_result)
-
-    X_pca = apca.transform(X)
-    assert_array_almost_equal(apca.explained_variance_,
-                              np.var(X_pca, ddof=1, axis=0))
-    assert_array_almost_equal(apca.explained_variance_, expected_result)
 
-    X_rpca = rpca.transform(X)
-    assert_array_almost_equal(rpca.explained_variance_,
-                              np.var(X_rpca, ddof=1, axis=0),
-                              decimal=1)
-    assert_array_almost_equal(rpca.explained_variance_,
-                              expected_result, decimal=1)
+@pytest.mark.parametrize("svd_solver", ['arpack', 'randomized'])
+def test_pca_singular_values_consistency(svd_solver):
+    rng = np.random.RandomState(0)
+    n_samples, n_features = 100, 80
+    X = rng.randn(n_samples, n_features)
 
-    # Same with correlated data
-    X = datasets.make_classification(n_samples, n_features,
-                                     n_informative=n_features-2,
-                                     random_state=rng)[0]
+    pca_full = PCA(n_components=2, svd_solver='full', random_state=rng)
+    pca_other = PCA(n_components=2, svd_solver=svd_solver, random_state=rng)
 
-    pca = PCA(n_components=2).fit(X)
-    rpca = PCA(n_components=2, svd_solver='randomized',
-               random_state=rng).fit(X)
-    assert_array_almost_equal(pca.explained_variance_ratio_,
-                              rpca.explained_variance_ratio_, 5)
+    pca_full.fit(X)
+    pca_other.fit(X)
 
+    assert_allclose(
+        pca_full.singular_values_, pca_other.singular_values_, atol=1e-1
```

Comment:
Actually, I think we should avoid using atol, except when comparing with 0,

cf https://github.com/scikit-learn/scikit-learn/pull/14178#issuecomment-505182824 and https://www.python.org/dev/peps/pep-0485/#absolute-tolerance-default

Reply:
Good point